### PR TITLE
this addresses issue #22 by using the key when no translation is found

### PR DESCRIPTION
**Pull request by mvolkmann**

### CODE REVIEW

**Comment by mvolkmann** [comment]
I signed it!

File: app-localize-behavior.html [hunk 1]
```diff
@@ -167,6 +179,20 @@
...
   },
 
+  _languageChanged() {
+    // Clear the message cache.
+    var proto = this.constructor.prototype;
+    proto.__localizationCache.messages = {};
+
+    for (const instance of appLocalizeBehaviorInstances) {
+      // Update the language of all instances to match this one.
+      instance.language = this.language;
```
Comment (notwaldorf) [inline app-localize-behavior.html]: Hmm, this seems unrelated to the problem you're fixing, and slightly way bigger -- this basically makes it impossible to keep elements that use this behaviour use different languages at the same time. I don't think setting a language is that complicated or verbose, so I'd lean towards not doing this (certainly not doing it in this PR)

File: app-localize-behavior.html [hunk 2]
```diff
@@ -194,6 +220,16 @@
     }
   },
 
+  ready() {
+    console.log('app-localize-behavior.html ready: this.useKeyIfMissing =', this.useKeyIfMissing);
```
Comment (notwaldorf) [inline app-localize-behavior.html]: No `console.log`s, please

File: app-localize-behavior.html [hunk 3]
```diff
@@ -194,6 +220,16 @@
...
+    // If the instance has the languageChanged method, call it.
+    // This allows it to set its initial state based on the language.
+    if (this.languageChanged) this.languageChanged();
```
Comment (notwaldorf) [inline app-localize-behavior.html]: If statements on a separate line, with `{}` please

File: app-localize-behavior.html (hunk 1 above)
Comment (mvolkmann) [inline app-localize-behavior.html]: Sorry, I didn't mean to push that. I intended for that to be a separate pull request for a separate issue. You can ignore this commit.

I would however like the option to have all instances use the same language. I think that will be common case for most apps.

File: app-localize-behavior.html [hunk 4]
```diff
@@ -210,20 +246,30 @@
 
       // Cache the key/value pairs for the same language, so that we don't
       // do extra work if we're just reusing strings across an application.
-      var messageKey = key + resources[language][key];
-      var msg = proto.__localizationCache.messages[messageKey];
+      var value = resources[language][key];
```
Comment (notwaldorf) [inline app-localize-behavior.html]: can we name this something better than value? perhaps `translatedValue`?

File: app-localize-behavior.html [hunk 5]
```diff
@@ -210,20 +246,30 @@
...
+        msg = proto.__localizationCache.messages[messageKey];
+      }
 
-      if (!msg) {
-        msg = new IntlMessageFormat(resources[language][key], language, formats);
-        proto.__localizationCache.messages[messageKey] = msg;
+      if (value && !msg) {
+        try {
+          msg = new IntlMessageFormat(value, language, formats);
+          if (msg) proto.__localizationCache.messages[messageKey] = msg;
```
Comment (notwaldorf) [inline app-localize-behavior.html]: `If` statements on a new line, with `{}` please

File: app-localize-behavior.html [hunk 6]
```diff
@@ -210,20 +246,30 @@
...
+        } catch (e) {
+          if (!this.useKeyIfMissing) throw e;
```
Comment (notwaldorf) [inline app-localize-behavior.html]: Does this throw? When does this throw? I think adding a comment about this would be nice.

Also, why not assign `msg` to `key` in this block, rather than doing another check in L270?

File: app-localize-behavior.html [hunk 7]
```diff
@@ -115,7 +116,9 @@
      * The language used for translation.
      */
     language: {
-      type: String
+      type: String,
+      value: Polymer.defaultLanguage,
```
Comment (notwaldorf) [inline app-localize-behavior.html]: Oof, I don't know about this. We set default values fairly differently -- the element that implements this behaviour can set a `value` for this attribute in its properties block. I don't think this is a good pattern

File: demo/index.html [hunk 8]
```diff
@@ -17,13 +17,20 @@
...
-    <link rel="import" href="x-local-translate.html">
 
     <!-- Intl polyfill -->
     <script src="../../intl/dist/Intl.min.js"></script>
     <script src="../../intl/locale-data/jsonp/en.js"></script>
+
+    <!-- Must import polymer before running next script. -->
+    <link rel="import" href="../../polymer/polymer.html">
+    <!-- Set this before importing components that use app-localize-behavior. -->
+    <script>Polymer.defaultLanguage = 'fr'</script>
```
Comment (notwaldorf) [inline demo/index.html]: As mentioned, I'm not super keen on this, as this is not how we set default values on elements

File: app-localize-behavior.html (hunk 1 above)
Comment (mvolkmann) [inline app-localize-behavior.html]: I will push a new version soon that removes the "language change" code.

File: app-localize-behavior.html (hunk 2 above)
Comment (mvolkmann) [inline app-localize-behavior.html]: Removed in next version.

File: demo/x-local-translate.html [hunk 9]
```diff
@@ -45,12 +45,6 @@
 
       properties: {
         /* Overriden from AppLocalizeBehavior */
-        language: {
-          value: 'en',
```
Comment (notwaldorf) [inline demo/x-local-translate.html]: This was already setting a default value of `en` on this element -- I'm not sure I understand why this didn't work for you. You could also do 
```
language: {
  value: Polymer.defaultLanguage
}
```
here in your app if you wanted (but note, not in the behaviour itself like you've done it)

File: app-localize-behavior.html (hunk 4 above)
Comment (mvolkmann) [inline app-localize-behavior.html]: trnaslatedValue and translatedMessage sound good. The name "msg" was already being used, but I'll change it to "translatedMessage".

**Comment by notwaldorf** [comment]
I think there's a bunch of problems with this PR right now -- I don't think I understand the reasoning behind changing how the default values are specified, nor the bit about assuming everyone has a `languageChanged` method on the element implementing this behaviour -- those both look like changes you want to use in _your_ element implementing the behaviour, not the behaviour itself.

The change in ` __computeLocalize` looks ok, but needs a bit more cleanup.

Let me know when this is ready for another look!

**Comment by mvolkmann** [comment]
Okay, I think this is ready for another look now. I removed all the changes related to support global language setting and simplified my changes for useKeyIfMissing.

File: app-localize-behavior.html [hunk 10]
```diff
@@ -115,7 +115,7 @@
      * The language used for translation.
      */
     language: {
-      type: String
+      type: String,
```
Comment (notwaldorf) [inline app-localize-behavior.html]: no trailing comma, please

File: demo/x-local-translate.html [hunk 11]
```diff
@@ -50,7 +50,6 @@
           type: String
         },
 
-        /* Overriden from AppLocalizeBehavior */
```
Comment (notwaldorf) [inline demo/x-local-translate.html]: I think this comment is still good :)

File: app-localize-behavior.html [hunk 12]
```diff
@@ -157,6 +157,15 @@
     },
 
     /**
+     * When there is no translation for a given key,
+     * the key will be used when this is true.
```
Comment (notwaldorf) [inline app-localize-behavior.html]: How about rephrasing to "If true, will use the provided key when the translation does not exist for that key"

**Comment by notwaldorf** [comment]
This looks good -- just a couple of nits left. Also, would you mind squashing your commits, please?

File: app-localize-behavior.html (hunk 10 above)
Comment (mvolkmann) [inline app-localize-behavior.html]: fixed in next push

File: demo/x-local-translate.html (hunk 11 above)
Comment (mvolkmann) [inline demo/x-local-translate.html]: The comment is added back in my next commit.

File: app-localize-behavior.html (hunk 12 above)
Comment (mvolkmann) [inline app-localize-behavior.html]: The comment is changed to your suggestion in my next push.

**Comment by mvolkmann** [comment]
I made the changes you suggested and squashed the commits. It was my first time squashing commits. Hope I did it correctly.

**Comment by mvolkmann** [comment]
Ignore this pull request. It has been superceded by https://github.com/PolymerElements/app-localize-behavior/pull/43
